posts/inside-rust/2023-11-10-parallel-rustc.md

```diff
@@ -0,0 +1,239 @@
+deadlocks that still occasionally affect multi-threaded mode. Parallel
+execution is harder to get right than serial execution, even in Rust. For this
+reason the parallel front-end also won't be shipped in Beta or Stable releases
+for some time.
+
+### Performance effects
+
+When the parallel front-end is run in single-threaded mode, compilation times
+are typically within 2% of the serial front-end, which should be barely
+noticeable.
+
+When the parallel front-end is run in multi-threaded mode with `-Z threads=8`,
+out [measurements on real-world
+code](https://github.com/rust-lang/compiler-team/issues/681) on show that
+compile times can be reduced by up to 50%, though the affects vary widely and
+depend greatly on the characteristics of the code being compiled and the build
+configuration. For example, dev builds are likely to see bigger improvements
+than release builds because release builds usually spend more time doing
+optimizations in the back-end. A small number of cases compile more slowly in
+multi-threaded mode than single-threaded mode.
+
+We recommend eight threads because this is the configuration we have tested the
+most and it is known to give good results. Values lower than eight will see
+smaller benefits. Values greater than eight will give diminishing returns and
+may even give worse performance.
+
+If a 50% improvement seems low when going from one to eight threads, recall
+from the explanation above that the front-end only accounts for part of compile
+times, and the back-end is already parallel. You can't beat [Amdahl's
+Law](https://en.wikipedia.org/wiki/Amdahl%27s_law).
+
+Memory usage can increase significantly in multi-threaded mode. This is
+unsurprising given that various tasks, each of which requires a certain amount
+of memory, are now executing in parallel. We have seen increases of up to 35%.
+
+### Correctness
+
+Reliability in single-threaded mode should be high.
+
+In multi-threaded mode there are some known bugs, including deadlocks. If
+compilation hangs, you have probably hit one of them.
+
+### Feedback
+
+If you have any problems with the parallel front-end, please [file an issue
+marked with the "WG-compiler-parallel"
+label](https://github.com/rust-lang/rust/labels/WG-compiler-parallel).
+That link also shows existing known problems.
+
+For more general feedback, please start a discussion on the [wg-parallel-rustc
+Zulip
+channel](https://rust-lang.zulipchat.com/#narrow/stream/187679-t-compiler.2Fwg-parallel-rustc).
+We are particularly interested to hear the performance effects on the code you
+care about.
+
+# Future work
+
+We are working to improve the performance of the parallel front-end. As the
+graphs above showed, there is room to improve the utilization of the threads in
+the front-end.
+
+We are also working to iron out the remaining bugs in multi-threaded mode. We
+aim to stabilize the `-Z threads` option and ship the parallel front-end in
+multi-threaded mode on Stable releases in 2024.
```

Comment:
Are we aiming to make n > 1 the default in 2024, or just n=1 the default on stable?

Reply:
n > 1, as far as I know. (That's what "multi-threaded mode" refers to.) That's the whole point :)

Reply:
Ah, I see. This makes sense.